### PR TITLE
Added Twilio as a tool

### PR DESCRIPTION
**Pull request by jayeshp19**
Implements https://github.com/phidatahq/phidata/issues/1334

image

PTAL @manthanguptaa @ysolanky

### CODE REVIEW

**Review by anuragts — CHANGES_REQUESTED** [review]

File: phi/tools/twilio.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+            api_key: Twilio API Key (alternative to auth_token)
+            api_secret: Twilio API Secret (alternative to auth_token)
+            region: Twilio region (e.g. 'au1')
+            edge: Twilio edge location (e.g. 'sydney')
+            debug: Enable debug logging
+        """
+        super().__init__(name="twilio")
+        self.account_sid = account_sid or getenv("TWILIO_ACCOUNT_SID")
+        self.auth_token = auth_token or getenv("TWILIO_AUTH_TOKEN")
+        self.api_key = api_key or getenv("TWILIO_API_KEY")
```
[the remark this answers is not in the document]
Comment (jayeshp19) [reply phi/tools/twilio.py:41]: On it

File: cookbook/tools/twilio_tools.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+Usage:
+- Set the following environment variables:
+    export TWILIO_ACCOUNT_SID="your_account_sid"
+    export TWILIO_AUTH_TOKEN="your_auth_token"
+
+- Or provide them when creating the TwilioTools instance
+"""
+
+twilio_tools = TwilioTools()
```
Comment (anuragts) [inline cookbook/tools/twilio_tools.py:20]: directly pass twilioTools in the agent
like this     
tools=[DuckDuckGo()],

File: phi/tools/twilio.py [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+from phi.tools import Toolkit
+from phi.utils.log import logger
+
+
+try:
+    from twilio.rest import Client
+    from twilio.base.exceptions import TwilioRestException
+    from twilio.twiml.voice_response import VoiceResponse
+except ImportError:
+    raise ImportError("`twilio` not installed.")
```
Comment (anuragts) [inline phi/tools/twilio.py:13]: can you also add the command to install the libraries like `pip install ....`

File: cookbook/tools/twilio_tools.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+agent = Agent(
```
Comment (anuragts) [inline cookbook/tools/twilio_tools.py:22]: Also define model here 
add this
    model=OpenAIChat(id="gpt-4o"),

**Review by anuragts — CHANGES_REQUESTED** [review]

File: cookbook/tools/twilio_tools.py (hunk 4 above)
Comment (anuragts) [inline cookbook/tools/twilio_tools.py:22]: also give an unique name to this agent
like this 
    name="Web Agent",

File: cookbook/tools/twilio_tools.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        """You can help users by:
+        - Sending SMS messages
+        - Making phone calls
+        - Checking message history
+        - Generating TwiML responses using twilio_tools
+        
+        Always confirm before sending messages or making calls."""
+    ],
+    tools=[twilio_tools],
+    show_tool_calls=True,
```
Comment (anuragts) [inline cookbook/tools/twilio_tools.py:33]: also add  markdown=True

**Review by anuragts — CHANGES_REQUESTED** [review]

File: cookbook/tools/twilio_tools.py (hunk 2 above)
Comment (jayeshp19) [reply cookbook/tools/twilio_tools.py]: done

File: phi/tools/twilio.py (hunk 3 above)
Comment (jayeshp19) [reply phi/tools/twilio.py]: added

File: phi/tools/twilio.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jayeshp19) [reply phi/tools/twilio.py:47]: Added necessary changes

Context:
Two authentication methods are supported:
        1. Account SID + Auth Token
        2. Account SID + API Key + API Secret

File: cookbook/tools/twilio_tools.py (hunk 4 above)
Comment (jayeshp19) [reply cookbook/tools/twilio_tools.py:22]: added

Comment (jayeshp19) [reply cookbook/tools/twilio_tools.py:22]: added

File: cookbook/tools/twilio_tools.py (hunk 5 above)
Comment (jayeshp19) [reply cookbook/tools/twilio_tools.py:35]: done

File: cookbook/tools/twilio_tools.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    instructions=[
+        """You can help users by:
+        - Sending SMS messages
+        - Making phone calls
+        - Checking message history
+        - Generating TwiML responses using twilio_tools
+        
+        Always confirm before sending messages or making calls."""
+    ],
+    model=OpenAIChat(id="gpt-4o-mini"),
```
Comment (anuragts) [inline cookbook/tools/twilio_tools.py:33]: use gpt-4o

File: phi/tools/twilio.py [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+        self.api_secret = api_secret or getenv("TWILIO_API_SECRET")
+        
+        # Optional region and edge
+        self.region = region or getenv("TWILIO_REGION")
+        self.edge = edge or getenv("TWILIO_EDGE")
+
+        # Validate required credentials
+        if not self.account_sid:
+            logger.error("TWILIO_ACCOUNT_SID not set. Please set the TWILIO_ACCOUNT_SID environment variable.")
+            raise ValueError("Twilio Account SID is required")
```
Comment (anuragts) [inline phi/tools/twilio.py:57]: don't raise valueError.

File: phi/tools/twilio.py [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+
+        # Initialize client based on provided authentication method
+        if self.api_key and self.api_secret:
+            # Method 2: API Key + Secret
+            logger.info("Initializing Twilio client with API Key authentication")
```
Comment (anuragts) [inline phi/tools/twilio.py:62]: remove this logger.info

File: phi/tools/twilio.py [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+            self.client = Client(
+                self.api_key,
+                self.api_secret,
+                self.account_sid,
+                region=self.region or None,
+                edge=self.edge or None,
+            )
+        elif self.auth_token:
+            # Method 1: Auth Token
+            logger.info("Initializing Twilio client with Auth Token authentication")
```
Comment (anuragts) [inline phi/tools/twilio.py:72]: remove this logger.info

File: phi/tools/twilio.py [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+                self.auth_token,
+                region=self.region or None,
+                edge=self.edge or None,
+            )
+        else:
+            logger.error(
+                "Neither (auth_token) nor (api_key and api_secret) provided. "
+                "Please set either TWILIO_AUTH_TOKEN or both TWILIO_API_KEY and TWILIO_API_SECRET environment variables."
+            )
+            raise ValueError("No valid authentication method provided")
```
Comment (anuragts) [inline phi/tools/twilio.py:84]: don't raise valueError.

File: phi/tools/twilio.py [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+
+
+        logger.info("Twilio client initialized")
```
Comment (anuragts) [inline phi/tools/twilio.py:87]: remove this

**Review by anuragts — CHANGES_REQUESTED** [review]
Can you also remove calls (voice), lets do text messages for now

File: cookbook/tools/twilio_tools.py (hunk 6 above)
Comment (jayeshp19) [reply cookbook/tools/twilio_tools.py]: done

**Comment by Ansub** [comment]
That's a solid PR @jayeshp19, Thank you for your contribution to improving phidata 🙌

Keep up the great work! Team will soon review the PR 😊

**Comment by jayeshp19** [comment]
> That's a solid PR @jayeshp19, Thank you for your contribution to improving phidata 🙌
> 
> Keep up the great work! Team will soon review the PR 😊

Thank you @Ansub Really appreciate it :)

**Review by anuragts — APPROVED** [review]
LGTM

**Review by anuragts — COMMENTED** [review]
LGTM

File: cookbook/tools/twilio_tools.py [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+agent = Agent(
+    name="Twilio Agent",
+    instructions=[
+        """You can help users by:
+        - Sending SMS messages
+        - Checking message history
+        - getting call details
+        
+        Always confirm before sending messages."""
```
Comment (anuragts) [inline cookbook/tools/twilio_tools.py:30]: Hey @jayeshp19  can you remove this line from instructions as in this example we can confirm.
`Always confirm before sending messages`

**Review by anuragts — CHANGES_REQUESTED** [review]

File: cookbook/tools/twilio_tools.py (hunk 12 above)
Comment (jayeshp19) [reply cookbook/tools/twilio_tools.py]: Hi @anuragts Thanks for pointing out. I've updated code.

File: cookbook/tools/twilio_tools.py [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+        - getting call details
+        """
+    ],
+    model=OpenAIChat(id="gpt-4o"),
+    tools=[TwilioTools()],
+    show_tool_calls=True,
+    markdown=True,
+)
+
+agent.print_response("Can you send an SMS saying 'Your package has arrived' to +1234567890?")
```
Comment (manthanguptaa) [inline cookbook/tools/twilio_tools.py:37]: I tested out your PR but there was this issue that there is no sender phone number and receiver number segregation. Doing this resolved the issue for me 
```
sender_phone_number = ""
receiver_phone_number = ""

agent.print_response(
    f"Can you send an SMS saying 'Your package has arrived' to {receiver_phone_number} from {sender_phone_number}?"
)
```

Comment (manthanguptaa) [reply cookbook/tools/twilio_tools.py:37]: Can you make this change so that I can approve your PR and merge it ASAP?

Comment (jayeshp19) [reply cookbook/tools/twilio_tools.py:37]: on it

**Comment by jayeshp19** [comment]
@manthanguptaa added suggested changes

**Review by manthanguptaa — APPROVED** [review]

**Pull request merged**